ci: separate location for customized builds

rail opened the pull request:
Previously, we published custom builds to the same bucket and docker repository as the `release-*` and `master` builds.

This PR uses separate GCS bucket and GAR repository for builds other than on `master` and `release-*` branches. It also removes unused dockerhub login call.

Additionally, customized builds do not push the tag to the `release-staging` repo and do not consume release verification testing resources.

Fixes: DEVINF-590
Epic: None
Release notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95332)
<!-- Reviewable:end -->

rail commented:
bors ping

rail committed:
ci: separate location for custom builds

Previously, we published custom builds to the same bucket and docker
repository as the release-* and master builds.

This PR uses separate GCS bucket and GAR repository for builds other
than on master and release-* branches. It also removes unused dockerhub
login call.

Additionally, customized builds do not push the tag to the
`release-staging` repo and do not consume release verification testing
resources.

Fixes: DEVINF-590
Epic: None
Release note: None

rail commented:
This looks fine. 
<img width="993" alt="image" src="https://github.com/cockroachdb/cockroach/assets/42865/e92cae23-8ad5-4730-8927-b3ee4b409190">
<img width="710" alt="image" src="https://github.com/cockroachdb/cockroach/assets/42865/7dc5c9ad-0f49-4952-8a8b-7c956f8c7997">

jlinder commented on build/teamcity/internal/release/process/make-and-publish-build-artifacts.sh:
```diff
@@ -126,5 +135,13 @@ cat << EOF
 Build ID: ${build_name}
 
+The binaries are available at:
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.linux-amd64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.linux-arm64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.darwin-10.9-amd64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.windows-6.2-amd64.zip
```
Let's have these print the location of the files for the location they were placed in.

```suggestion
  https://storage.googleapis.com/$gcs_bucket/cockroach-$build_name.linux-amd64.tgz
  https://storage.googleapis.com/$gcs_bucket/cockroach-$build_name.linux-arm64.tgz
  https://storage.googleapis.com/$gcs_bucket/cockroach-$build_name.darwin-10.9-amd64.tgz
  https://storage.googleapis.com/$gcs_bucket/cockroach-$build_name.windows-6.2-amd64.zip
```

jlinder commented on build/teamcity/internal/release/process/make-and-publish-build-artifacts.sh:
```diff
@@ -126,5 +135,13 @@ cat << EOF
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.windows-6.2-amd64.zip
+
+Pull the docker image by:
+  docker pull us-docker.pkg.dev/cockroach-cloud-images/cockroachdb-customized/cockroach-customized:$build_name
```
Let's have this print the repository where the image was pushed to.

```suggestion
  docker pull $gcr_repository:$build_name
```

jlinder commented on build/teamcity/internal/release/process/make-and-publish-build-tagging.sh:
```diff
@@ -28,51 +32,58 @@ fi
 Build ID: ${build_name}
 
+The binaries are available at:
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.linux-amd64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.linux-arm64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.darwin-10.9-amd64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.windows-6.2-amd64.zip
```
How about making these point at the right place for customized and dry run builds?

jlinder commented on build/teamcity/internal/release/process/make-and-publish-build-tagging.sh:
```diff
@@ -28,51 +32,58 @@ fi
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$build_name.windows-6.2-amd64.zip
+
+Pull the docker image by:
+  docker pull us-docker.pkg.dev/cockroach-cloud-images/cockroachdb-customized/cockroach-customized:$build_name
```
How about making this point at the right place for customized and dry run builds?

jlinder commented on scripts/tag-custom-build.sh:
```diff
@@ -75,4 +75,14 @@ Here is where the build run should show up in TeamCity for the tag:
 Build ID: $ID
+
+The binaries will be available at:
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$ID.linux-amd64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$ID.linux-arm64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$ID.darwin-10.9-amd64.tgz
+  https://storage.googleapis.com/cockroach-builds-artifacts-prod/cockroach-$ID.windows-6.2-amd64.zip
```
```suggestion
  https://storage.googleapis.com/cockroach-customized-builds-artifacts-prod/cockroach-$ID.linux-amd64.tgz
  https://storage.googleapis.com/cockroach-customized-builds-artifacts-prod/cockroach-$ID.linux-arm64.tgz
  https://storage.googleapis.com/cockroach-customized-builds-artifacts-prod/cockroach-$ID.darwin-10.9-amd64.tgz
  https://storage.googleapis.com/cockroach-customized-builds-artifacts-prod/cockroach-$ID.windows-6.2-amd64.zip
```

jlinder reviewed (commented):
A few suggestions. Otherwise, looks great!

jlinder reviewed (commented):
A few questions.

jlinder commented on build/teamcity/internal/release/process/make-and-publish-build-tagging.sh:
```diff
@@ -28,51 +37,59 @@ fi
+  build_name:          $build_name
+  release_branch:      $release_branch
+  is_customized_build: $is_customized_build
+  gcr_repository:      $gcr_repository
```
One last question, why not add the `gcs_bucket` to the output?

```suggestion
  gcs_bucket:          $gcs_bucket
  gcr_repository:      $gcr_repository
```

rail replied:
I'd say it a cargo cult thing. :D We print the full URLs at the end in any case. This is more like a debugging output IMO. I can add the bucket name if you have a strong opinion.

jlinder replied:
Not a strong opinion. It's good to go. :shipit:

jlinder reviewed (approved):
One last question. Good either way.

rail commented:
bors r=jlinder

rail commented:
blathers backport release-22.2
